tests/ship_restart_test.py

```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python3
+
+import os
+import tempfile
+import shutil
+import signal
+
+from TestHarness import Cluster, TestHelper, Utils, WalletMgr
+
+###############################################################################
+# ship_restart_test
+# 
+# This test verifies SHiP shuts down gracefully or recovers when restarting
+# with various scenarios of corrupted log and/or index files.
+# 
+###############################################################################
+
+Print=Utils.Print
+
+args = TestHelper.parse_args({"--dump-error-details","--keep-logs","-v","--leave-running","--unshared"})
+
+Utils.Debug=args.v
+cluster=Cluster(unshared=args.unshared, keepRunning=args.leave_running, keepLogs=args.keep_logs)
+dumpErrorDetails=args.dump_error_details
+walletPort=TestHelper.DEFAULT_WALLET_PORT
+
+totalProducerNodes=1
+totalNonProducerNodes=1 # for SHiP node
+totalNodes=totalProducerNodes+totalNonProducerNodes
+
+walletMgr=WalletMgr(True, port=walletPort)
+testSuccessful=False
+
+prodNodeId = 0
+shipNodeId = 1
+
+origStateHistoryLog   = ""
+stateHistoryLog       = ""
+origStateHistoryIndex = ""
+stateHistoryIndex     = ""
+
+# Verifies that SHiP should fail to restart with a corrupted first entry header
+def corruptedHeaderTest(pos, corruptedValue, shipNode):
+    # restore log and index
+    shutil.copyfile(origStateHistoryLog, stateHistoryLog)
+    shutil.copyfile(origStateHistoryIndex, stateHistoryIndex)
+
+    with open(stateHistoryLog, 'rb+') as f: # opened as binary file
+        f.seek(pos) # seek to the position to corrupt
+        f.write(corruptedValue) # corrupt it
+
+    isRelaunchSuccess = shipNode.relaunch()
+    assert not isRelaunchSuccess, "SHiP node should have failed to relaunch"
+
+try:
+    TestHelper.printSystemInfo("BEGIN")
+
+    cluster.setWalletMgr(walletMgr)
+    Print("Stand up cluster")
+
+    specificExtraNodeosArgs={}
+    specificExtraNodeosArgs[shipNodeId]="--plugin eosio::state_history_plugin --trace-history --chain-state-history --finality-data-history --state-history-stride 200 --plugin eosio::net_api_plugin --plugin eosio::producer_api_plugin"
```

Comment:
Does the chain ever make it past block 200 to exercise anything other then the simple "flat log" case? But I suppose there is probably not much that can be done for a corrupted non-head split log anyways -- fixing it just makes a hole that isn't allowed.

I do wonder if we should do some additional corruption tests with pruned logs. But those are even less repairable.

Reply:
> Does the chain ever make it past block 200 to exercise anything other then the simple "flat log" case? But I suppose there is probably not much that can be done for a corrupted non-head split log anyways -- fixing it just makes a hole that isn't allowed.
> 
> I do wonder if we should do some additional corruption tests with pruned logs. But those are even less repairable.

Thanks.  I created https://github.com/AntelopeIO/spring/issues/1011 to track additional tests for corrupted pruned and split tests. I think they should be done as a part of the bigger initiative to revisit how corrupted logs are handled (repairing at the startup or offline via `spring-util`.

I am merging the PR in current form so we have some basic tests.